modules/openapi-generator-gradle-plugin/src/main/kotlin/org/openapitools/generator/gradle/plugin/tasks/GenerateTask.kt

```diff
@@ -500,6 +500,8 @@ open class GenerateTask : DefaultTask() {
     @Suppress("unused")
     @TaskAction
     fun doWork() {
+        project.delete(outputDir)
```

Comment:
Yes that is true. But this should be the expected behavior of a Gradle task where generated code should never be modified.

Reply:
I feel we may be thinking of two separate use cases. 

The first use case is where sources are generated to the project's `build` directory as part of the build to use within the project. These files should never be directly modified as they live in the `build` directory which is used to store untracked files, artifacts, etc. Running a Gradle `clean`removes these files and re-running the build re-generates them. Tests are included as a completely separate source set and are tracked by version control. This is a very "ephemeral" approach code generation. I believe @sorin-florea's changes are to better support this particular use case.

The second use case is generating code which _is_ tracked by version control and _is_ committed as part of the project. It would be generated to some arbitrary directory, not to Gradle's `build` directory. It's code that is meant to be long lasting. In this case you do not want the directory to be removed when re-generating code. @wing328 I believe this is the use case you are describing.

I'm not sure which use case is more common by users of the Gradle plugin. I would imagine it is the former use case but that is simply an assumption. It is the only way I have seen it used. 

@wing328 do you feel the second use case is something the Gradle plugin should support? If so, something we can do is create a boolean task input for opting-in/out of clearing the output directory as part of task execution. Preferably this would be an opt-out where by default we delete the output directory. This aligns with the fact that [the default output directory is within the `build` directory](https://github.com/OpenAPITools/openapi-generator/blob/4667b7e4719ae4af2822ac451acbd13a2c7ef04c/modules/openapi-generator-gradle-plugin/src/main/kotlin/org/openapitools/generator/gradle/plugin/OpenApiGeneratorPlugin.kt#L63) which is inherently ephemeral. This certainly supports the first use case and would make the plugin "correct by default" in my eyes.

@wing328 at some point I would like to contribute a sample demonstrating how to properly implement the first use case. The [current sample](https://github.com/OpenAPITools/openapi-generator/blob/4667b7e4719ae4af2822ac451acbd13a2c7ef04c/modules/openapi-generator-gradle-plugin/samples/local-spec/build.gradle) is close, but it doesn't include the generated sources as a source set so you can't reference the generated sources within the main or test source sets.

Lastly, how to best handle generated sources is something we're looking to add to our documentation.

Reply:
> These files should never be directly modified as they live in the build directory which is used to store untracked files, artifacts, etc

I agree there are use cases in which users want to regenerate the auto-generated code by purging the output folder to start with as those code is never meant to be modified manually "by design".

From my experience working on this project, there are always users not following "design". This is not to say they are doing the right thing but just something to consider to avoid surprise to these users (especially new users) when using openapi-generator.

> The second use case is generating code which is tracked by version control and is committed as part of the project

That's the usual assumption that a user will track the files using version control system (VCS) but there are always edge cases...

> If so, something we can do is create a boolean task input for opting-in/out of clearing the output directory as part of task execution

opting out by default (i.e. output folder won't be purge by default) is much better IMO, especially to new users using openapi-generator.

One thing that I would like to find out more is why not simply let the users remove/purge the folders manually if needed via command line for example before using the Gradle plugin?

For example, if users would prefer the output conforms to certain code style, we encourage them to run a code formatter/linter post code-generation (instead of embedding the operation/logic into openapi-generator itself). This is not to say code formatting and purging output folder pre-code generation is the same thing but I hope you see my point.

Reply:
I agree with Eric's two different use-cases.

IMHO, if a source generator is part of a build, then it's output should not be committed, only the necessary inputs. Otherwise, problems arise when the inputs are changed, but the generated sources have been manually modified. In this case, you either won't get the updates, or your modifications get overwritten, both cases are sub optimal. I can't think of a source generator who doesn't follow this pattern.

If we want to support both use-cases in a single task, then a boolean flag would be a good solution. Alternatively, we could make it "smart" by detecting if the output directory is within the `build` directory and then treating it as intended to be regenerated. Although, this behavior could be surprising to users. We could also have two tasks, one intended to be invoked only manually, the other to be added to an automated build.

Reply:
@wing328 I added a boolean flag.

Reply:
thanks for adding that and defaulting it to false so that the output directory is not removed before code generation (backward-compatible).